src/task_handle.rs

```diff
@@ -0,0 +1,138 @@
+pub use self::tk::TaskHandle;
+
+use crate::error::{ZmqError, ZmqResult};
+
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum TaskError {
+    #[error("Internal error: {0}")]
+    Internal(#[from] Box<ZmqError>),
+    #[error("Task panicked")]
+    Panic,
+    #[error("Task cancelled")]
+    Cancelled,
+}
+impl From<tokio::task::JoinError> for TaskError {
+    fn from(err: tokio::task::JoinError) -> Self {
+        if err.is_panic() {
+            TaskError::Panic
+        } else {
+            debug_assert!(err.is_cancelled());
+            TaskError::Cancelled
+        }
+    }
+}
+
+mod tk {
+    use super::*;
+
+    pub struct TaskHandle<T> {
+        // Using options to allow us to move resource without consuming `self`
+        stop_channel: Option<futures::channel::oneshot::Sender<()>>,
```

Comment:
What's the point in wrapping this into Option? Can we ever construct/use this struct with None values?

Reply:
Originally I did this because I needed the shutdown function to take an `&mut Self`, but it takes a `Self` now. I can probably change these back to the option-less version.

Reply:
Imho it's better not to use Option when it's not needed. This makes this structure more reliable

Reply:
Oh I remember now...

```
error[E0509]: cannot move out of type `TaskHandle<T>`, which implements the `Drop` trait
  --> src/task_handle.rs:54:32
   |
54 |             let stop_channel = self.stop_channel; //.take().unwrap();
   |                                ^^^^^^^^^^^^^^^^^
   |                                |
   |                                cannot move out of here
   |                                move occurs because `self.stop_channel` has type `futures::channel::oneshot::Sender<()>`, which does not implement the `Copy` trait
   |                                help: consider borrowing here: `&self.stop_channel`
```

I implemented Drop so that we can get warnings when we dropped the contents during panics without shutting down. For that reason, I can't actually destructure the struct in a way that lets me consume its fields. To get around it, I used `Option` to allow me to `take()` the contents and assume ownership over them.

Reply:
Do we really need this log?
Even in case of panic stop_channel will be dropped and this will cause task to stop (just without explicit wait).

Reply:
So we have `shutdown()` which we can call whenever we need an explicit way to block until the task is finished - in the case of the `ipc` socket logic for example, shutdown is important to make sure resources are released before doing further operations to avoid a race condition that would leak a file descriptor and prevent rebinding to the same socket file. 

However, suppose we panic before calling `shutdown()`. In this case, the task will eventually cancel due to its `drop()` implementation, but we have potentially introduced a race condition somewhere else in the code (since the parent task was supposed to `await` until the child task completed). For debugging purposes therefore, it was helpful for me when I was doing to IPC stuff to see if we ever drop a `TaskHandle` while panicking.

Reply:
> However, suppose we panic before calling shutdown(). In this case, the task will eventually cancel due to its drop() implementation, but we have potentially introduced a race condition somewhere else in the code (since the parent task was supposed to await until the child task completed).

From what I know we are not expected to provide any kind of guarantees in case of panic. In most cases this will simply stop current process and exit.

I don't think that it's a good idea to change code structure just for debug purposes (cause it introduces some additional complexity without any good reason). Without Option you could also remove is_shutting_down property (cause you can only shutdown this Handle once). Overall it would make it simpler and easier to understand.